### PR TITLE
#106 Have some basic launch data available to UI

### LINKED ISSUE #106: Sending and holding LTI Launch information in the Frontend and Backend respectively
The LTI launch information needs to be remembered in the Backend past the LTI launch. Currently the idea is to put it in the request/response session. Also send certain information to Frontend so that the it load features based on the canvas course role etc.

### PR DESCRIPTION
**Pull request by pushyamig**
Fixes #106 

This PR builds upon on the work done by @ssciolla for integrating Express session library #22. I am pulling certain information from Launch that is will be useful for hold during a user session. I am also sending some information to Frontend as well. I am fulfilling the data needs at this point of time. If there are more we could easily aded

### CODE REVIEW

File: config/lti_dev_key_sample.json [hunk 1]
```diff
@@ -24,7 +24,8 @@
   "custom_fields": {
     "roles": "$Canvas.membership.roles",
     "is_root_account_admin": "$Canvas.user.isRootAccountAdmin",
-    "login_id": "$Canvas.user.loginId"
+    "login_id": "$Canvas.user.loginId",
+    "course_id": "$Canvas.course.id"
```
Comment (ssciolla) [inline config/lti_dev_key_sample.json:29]:
```suggestion
```

File: ccm_web/server/localTypes/express-session/index.d.ts [hunk 2]
```diff
@@ -4,6 +4,9 @@ declare module 'express-session' {
   interface CustomData {
     ltiKey: string
     userLoginId: string
+    courseId: number
+    roles: string[]
```
Comment (ssciolla) [inline ccm_web/server/localTypes/express-session/index.d.ts:8]: We could do a little more research on what role types are possible and be more specific with the type. I'm thinking something like this, or an enum:

```
type CourseRoleType = 'TeacherEnrollment' | 'StudentEnrollment' ...
```

Also, if these roles are context specific, maybe we should the data shape here somewhat, maybe do something like this:
```
course: {
  id: number
  roles: CourseRoleType[]
}
```

File: ccm_web/server/src/api/api.service.ts [hunk 3]
```diff
@@ -10,9 +11,12 @@ export class APIService {
     }
   }
 
-  getGlobals (): Globals {
+  getGlobals (data: SessionData): Globals {
```
Comment (ssciolla) [inline ccm_web/server/src/api/api.service.ts:14]: Nit-pick: I think I'd use `sessionData` or `session` for the parameter name for clarity.

File: ccm_web/server/src/lti/lti.service.ts [hunk 4]
```diff
@@ -58,10 +58,14 @@ export class LTIService implements BeforeApplicationShutdown {
     provider.onConnect(async (token: IdToken, req: Request, res: Response) => {
       logger.debug(`The LTI launch was successful! User info: ${JSON.stringify(token.userInfo)}`)
       const customLTIVariables = token.platformContext.custom
-      if (customLTIVariables.login_id === undefined) {
+      if (customLTIVariables.login_id === undefined || customLTIVariables.course_id === undefined ||
+        customLTIVariables.roles === undefined || customLTIVariables.is_root_account_admin === undefined) {
```
Comment (ssciolla) [inline ccm_web/server/src/lti/lti.service.ts:62]: Not sure how many more properties there will be, but if this gets much longer, we might want to think about a more succinct validation strategy, using a schema or something. This looks okay for now.

File: ccm_web/server/src/api/api.controller.ts [hunk 5]
```diff
@@ -15,7 +17,7 @@ export class APIController {
   }
 
   @Get('globals')
-  getGlobals (): Globals {
-    return this.apiService.getGlobals()
+  getGlobals (@Req() req: Request, @Session() session: SessionData): Globals {
```
Comment (ssciolla) [inline ccm_web/server/src/api/api.controller.ts:20]: Do we need `Req` or `Request` yet in this file, if you're just going to use `session`?

**Review by ssciolla — COMMENTED** [review]
Great! Thanks for doing this. I left a couple thoughts, suggestions, but I think this looks good. I'm going to test and then I'll approve. Feel free to incorporate my suggestions or not.

**Comment by ssciolla** [comment]
@pushyamig, maybe we should go ahead and update the `Globals` interface in the client, too? `ccm_web/client/src/models/models.ts`

File: ccm_web/server/src/api/api.controller.ts (hunk 5 above)
Comment (pushyamig) [reply ccm_web/server/src/api/api.controller.ts]: Removed the Request.

File: ccm_web/server/src/api/api.service.ts (hunk 3 above)
Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts]: Make sense to do it like that

File: ccm_web/server/src/lti/lti.service.ts (hunk 4 above)
Comment (pushyamig) [reply ccm_web/server/src/lti/lti.service.ts:62]: Yeah...

File: config/lti_dev_key_sample.json (hunk 1 above)
Comment (pushyamig) [reply config/lti_dev_key_sample.json]: Done

**Comment by pushyamig** [comment]
> @pushyamig, maybe we should go ahead and update the Globals interface in the client, too? ccm_web/client/src/models/models.ts

Done [`33a6e80` (#114)](https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/114/commits/33a6e803d2101f51ff530cd195c6d31f83184724)

File: ccm_web/server/localTypes/express-session/index.d.ts (hunk 2 above)
Comment (chrisrrowland) [reply ccm_web/server/localTypes/express-session/index.d.ts:8]: I would think we want to at expect multiple roles?

```
if (roles.contains(TEACHER_ROLE)) {
   userFeatures.concat(TEACHER_FEATURES)
}
if (roles.contains(ADMIN_ROLE) {
  userFeatures.concat(ADMIN_FEATURES)
}
...
   
```

Comment (pushyamig) [reply ccm_web/server/localTypes/express-session/index.d.ts:8]: Yes, a user can have multiple roles but if he is admin then we just show then all features

Comment (ssciolla) [reply ccm_web/server/localTypes/express-session/index.d.ts:8]: Okay, @pushyamig, so you want to kick the can on specific types down the road? I'm okay with that. What do you think about the `course` nesting?

File: config/lti_dev_key_sample.json [hunk 6]
```diff
@@ -24,9 +24,9 @@
   "custom_fields": {
     "roles": "$Canvas.membership.roles",
     "is_root_account_admin": "$Canvas.user.isRootAccountAdmin",
-    "login_id": "$Canvas.user.loginId"
-    
-  },
+    "login_id": "$Canvas.user.loginId",
+    "course_id": "$Canvas.course.id"
+    },
```
Comment (ssciolla) [inline config/lti_dev_key_sample.json:29]:
```suggestion
  },
```

**Review by ssciolla — APPROVED** [review]
I tested locally, seems to be working well! I'm approving. You can use your discretion to resolve anything pending/create new issues/etc.

I'll rebase my #99 branch after you merge this.

File: ccm_web/server/localTypes/express-session/index.d.ts (hunk 2 above)
Comment (pushyamig) [reply ccm_web/server/localTypes/express-session/index.d.ts]: @ssciolla I changed course object you [suggested](https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/114/commits/8b7d6d2f7faeefecfef6f870de796d84cb321347). It was better way to do it

**Comment by pushyamig** [comment]
I resolved the issue from @ssciolla suggestion and what I felt was important to do

**Comment by ssciolla** [comment]
@pushyamig, cool, LGTM!

**Pull request merged**